generator/tests/test_examples/run_examples_tests.py

```diff
@@ -22,117 +21,208 @@
 ##############################################################################
 # Specific generation functions
 
+def set_up(filename):
+    """
+    Parse an XML file and get the big dictionary structure of information
+    from it.
 
-def generate_example(filename):
+    :param filename: the file to parse
+    :return: the big dictionary structure.
+    """
     parser = ParseXML.ParseXML(filename)
     ob = parser.parse_deviser_xml()
-    os.chdir('./temp')
+    os.chdir(os.path.normpath('./temp'))
     if not os.path.exists(ob['name']):
         os.mkdir(ob['name'])
-    os.chdir('./{0}'.format(ob['name']))
+    os.chdir(os.path.normpath('./{0}'.format(ob['name'])))
+    return ob
+
+
+def generate_example(filename):
+    """
+    Generate a C++ example.
+
+    :param filename: XML file to parse
+    :return: package name, e.g. 'dyn'
+    """
+    ob = set_up(filename)
     all_files = CppExampleFile.CppExampleFile(ob)
     all_files.write_file()
-    os.chdir('../../.')
+    os.chdir(os.path.normpath('../../.'))
     return ob['name']
 
 
 def generate_xml(filename):
-    parser = ParseXML.ParseXML(filename)
-    ob = parser.parse_deviser_xml()
-    os.chdir('./temp')
-    if not os.path.exists(ob['name']):
-        os.mkdir(ob['name'])
-    os.chdir('./{0}'.format(ob['name']))
+    """
+    Generate an XML example.
+
+    :param filename: XML file to parse
+    :return: package name, e.g. 'dyn'
+    """
+    ob = set_up(filename)
     all_files = ValidationXMLFiles.ValidationXMLFiles(ob)
     all_files.write_file('test_xml')
-    os.chdir('../../.')
+    os.chdir(os.path.normpath('../../.'))
     return ob['name']
 
 
 def generate_xml_fails(filename):
-    parser = ParseXML.ParseXML(filename)
-    ob = parser.parse_deviser_xml()
-    os.chdir('./temp')
-    if not os.path.exists(ob['name']):
-        os.mkdir(ob['name'])
-    os.chdir('./{0}'.format(ob['name']))
+    """
+    Generate a failure example?????? TODO unsure
+
+    :param filename: XML file to parse
+    :return: package name, e.g. 'dyn'
+    """
+    ob = set_up(filename)
     all_files = ValidationXMLFiles.ValidationXMLFiles(ob)
     all_files.write_all_files()
-    os.chdir('../../.')
+    os.chdir(os.path.normpath('../../.'))
     return ob['name']
 
+
 def generate_some_xml_fails(filename, start, stop, number=-1):
-    parser = ParseXML.ParseXML(filename)
-    ob = parser.parse_deviser_xml()
-    os.chdir('./temp')
-    if not os.path.exists(ob['name']):
-        os.mkdir(ob['name'])
-    os.chdir('./{0}'.format(ob['name']))
+    """
+    Generate
+
+    :param filename: XML file to parse
+    :param start:
+    :param stop:
+    :param number: number of files to write? TODO not sure.
+    :return: package name, e.g. 'dyn'
+
+    TODO Sorry Sarah I'm a bit confused by this one!
```

Comment:
his was a work in progreess and is not included in overall test run

Reply:
ok, comment updated